src/rpc/state_api.rs

```diff
@@ -804,6 +804,60 @@ pub(in crate::rpc) async fn state_vm_circulating_supply_internal<
     )?))
 }
 
+/// Looks back and returns all messages with a matching to or from address, stopping at the given height.
+pub(in crate::rpc) async fn state_list_messages<DB: Blockstore + Send + Sync + 'static>(
+    data: Data<RPCState<DB>>,
+    Params(LotusJson((from_to, tsk, max_height))): Params<
+        LotusJson<(MessageFilter, TipsetKeys, i64)>,
+    >,
+) -> Result<LotusJson<Vec<Cid>>, JsonRpcError> {
+    let ts = data.chain_store.load_required_tipset(&tsk)?;
+
+    if from_to.from == None && from_to.to == None {
+        return Err("must specify at least To or From in message filter".into());
+    } else if let Some(to) = from_to.to {
```

Comment:
what happens if we have both `to` and `from` specified?

Reply:
if we need to verify both, the correct way of doing that would be getting rid of the `else if` blocks and just doing:

```if let Some(to) ..```

```if let Some(from)..```

Reply:
I was following lotus logic since the results need to mathc lotus result:

https://github.com/filecoin-project/lotus/blob/9b4a3ba23dd4e8c8255754e48ab3f134ec8860ef/node/impl/full/state.go#L1175-L1176

but I agree your suggestion make more sense.

Reply:
Since this is following Lotus logic, which looks buggy to me, please keep following it.
But please mention this in a comment and provide this link to justify.